### PR TITLE
Get channel map and send to template

### PR DESCRIPTION
**Pull request by tbille**
# Summary

Send channel map to template on snap details page

# QA 

- `./run`
- 0.0.0.0:8004/toto
- should work

This is the structure sent to the template: https://pastebin.canonical.com/p/d4qs6fMRZJ/

### CODE REVIEW

**Comment by bartaz** [comment]
We want to display a release date for each channel, is it possible with current data from API or do we need to request an API change for that?

Other thing (could be done separtely) would be to change the structure of data a bit.
In UI we are displaying channel map by architecture and by track, so it would be nice to have the data structure reflect that.
Currently it's a flat list with `architecture` and `track` fields.

Ideally structure of the data would be:

```
channel_map = {
   'amd64': {
       'latest': [
           { channel: 'stable', release: 1, version: '1.0', ... }
           { channel: 'candidate', release: 2, version: '1.1', ... }
           ...
       ]
   }
   'armhf': {
       'latest': [ ... ]
       'othertrack': [ ... ]
   }
}
```

**Comment by tbille** [comment]
@bartaz I can update the structure of the json sent. But it doesn't seem that we can get the release date. This is something that should be asked to the store team.

**Comment by bartaz** [comment]
@tbille Yes, I think it would be helpful to have data structured in a way that will make it easier to access it in template/front-end.

And for the date - we can do a first implementation without it I think, but we should request this to be added to API.

**Comment by tbille** [comment]
@bartaz the channel maps sent to the template is as you asked for. Check the documentation added in the function added in `logic.py`

File: modules/public/logic.py [hunk 1]
```diff
@@ -281,3 +281,44 @@ def external(attrs, new=False):
         formatted_paragraphs.append(paragraph.replace('\n', '<br />'))
 
     return formatted_paragraphs
+
+
+def convert_channel_maps(channel_maps_list):
```
Comment (bartaz) [inline modules/public/logic.py]: No tests for this?

File: modules/public/logic.py [hunk 2]
```diff
@@ -281,3 +281,44 @@ def external(attrs, new=False):
...
+
+    :param channel_maps_list: The channel maps list returned by the API
+
+    :returns: The channel maps reshaped
+    """
+    channel_maps = {}
+    for channel_map in channel_maps_list:
+        arch = channel_map['architecture']
+        track = channel_map['track']
+        channel_maps[arch] = {}
```
Comment (bartaz) [inline modules/public/logic.py]: Does this work properly?
If there are 2 tracks for same architecture the second one will remove the first one if I read the code correctly.

Comment (bartaz) [reply modules/public/logic.py]: I QA'ed it and it doesn't work for example for data like this:

```
    channel_map = [
        {
            'architecture': 'arch',
            'map': [
                {'info': 'release', 'channel': 'stable'},
                {'info': 'release', 'channel': 'beta'}
            ],
            'track': 'latest'
        },
        {
            'architecture': 'arch',
            'map': [
              {'info': 'release', 'channel': 'stable'},
              {'info': 'release', 'channel': 'beta'}
            ],
            'track': 'track2'
        },
        ]
```

File: modules/public/logic.py (hunk 1 above)
Comment (tbille) [reply modules/public/logic.py]: tests added

File: modules/public/logic.py (hunk 2 above)
Comment (tbille) [reply modules/public/logic.py]: done

**Pull request merged**